pkg/controller/machineset.go

```diff
@@ -323,22 +324,39 @@ func (c *controller) manageReplicas(allMachines []*v1alpha1.Machine, machineSet
 		return nil
 	}
 
-	var activeMachines, staleMachines []*v1alpha1.Machine
+	var activeMachines, runningMachines, staleMachines []*v1alpha1.Machine
 	for _, machine := range allMachines {
 		if IsMachineActive(machine) {
-			//klog.Info("Active machine: ", machine.Name)
 			activeMachines = append(activeMachines, machine)
+
+			if IsMachineRunning(machine) {
+				runningMachines = append(runningMachines, machine)
+			}
 		} else if IsMachineFailed(machine) {
 			staleMachines = append(staleMachines, machine)
 		}
 	}
 
+	machineDeletionWindow := c.calculateStaleMachineDeletionWindow(activeMachines, runningMachines, staleMachines, machineSet)
+
 	if len(staleMachines) >= 1 {
 		klog.V(2).Infof("Deleting stale machines")
 	}
-	c.terminateMachines(staleMachines, machineSet)
+	deletedStaleMachines, _ := c.terminateStaleMachines(staleMachines, machineDeletionWindow, machineSet)
+	staleMachinesLeft := len(staleMachines) - deletedStaleMachines
 
 	diff := len(activeMachines) - int(machineSet.Spec.Replicas)
+	if staleMachinesLeft > 0 {
+		// Count the leftover stale machines against .spec.replicas to prevent sudden surge in machines which can happen if large number of stale machines are left over
+		diff -= staleMachinesLeft
```

Comment:
But that won't work, because it'll panic in `getMachinesToDelete()` since diff will be bigger than activeMachines count.

It's my fault for moving this calculation outside of the `if` branch. [This](https://github.com/gardener/machine-controller-manager/pull/482/files#diff-16f7b87f40829e0c81cb6c486f3997ccR361-R364) condition should suffice.